<a href="https://colab.research.google.com/github/aydanaghabayli/PolygonShape/blob/main/Copy_of_Exercise_3_1_Pandas_Time_Series_I.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science for Construction, Architecture and Engineering

### Week 3 - Construction - Pandas Time-Series Data

### Exercise Set 3.1 - Pandas Time-Series I

This set of exercises is the first of two sets for Week 3. These exercises are designed to give you some hands on experience in trying out the concepts from the first set of videos for this week.

There is an introduction video available to explain the process of using this exercise video.

The lead developer of this exercise is James Zhan Sicheng, a Ph.D. student at NUS and a teaching assistant of this course. 

# Exercise 03.1

There are three exercises in this set, focusing on the use of the [Pandas](https://pandas.pydata.org/) functions introduced in the videos: read, resample, truncate and plot.

### Please note that you will need to use the resultant output from Exercise 03.1.3 for the Quiz 3.1.

#### Setup the environment
The cell below is used to load the necessary packages for this exercise and setup the working directory. You will be asked to give the google drive authorization using through a given link.

In [ ]:
import pandas as pd
from google.colab import drive
import os
import numpy as np

In [ ]:
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In order to do the following exercises, we will go ahead and load the `meta` data file. Please note that we're using the `os` library again to navigate our file struture to *change the directory* to find the right files.

**Important Note** - if your file structure is different from what you see below, then you will get an error. You will need to change your file locations *or* update the file path

In [ ]:
# read meta data file as pandas dataframe and assign to the corresponding variables
os.chdir("/content/gdrive/My Drive/EDX Data Science for Construction, Architecture and Engineering/3 - Construction - Pandas Fundamentals/")
metadata = pd.read_csv('all_buildings_meta_data.csv')

### Exercise 03.1.1 - Loading Meter Data from Specific Files

Sometimes you need to grab and process data only from certain files to analyze. You want to first filter those buildings that have certain attributes.


Use `metadata` to find the specific building that meet the description and retrieve the required information. This code:
- Use the `.loc` function to identify the building located in **America/Phoenix**, having **College Laboratory** as `primaryspaceUsage` and having floor area larger than **30000** `sqm`
- print out the `uid`, `nickname`, `newweatherfilename` and `yearbuilt` of the building

In [ ]:
# Retrieve the required information according to the given conditions
metadata2 = metadata.loc[lambda metadata: metadata['timezone'] == 'America/Phoenix']
metadata3 = metadata2.loc[lambda metadata2: metadata2['primaryspaceusage'] == 'College Laboratory']
metadata4 = metadata3.loc[lambda metadata3: metadata3['sqm'] > 30000]
metadata_filtered = metadata4[['uid', 'newweatherfilename', 'yearbuilt']]
metadata_filtered


uid newweatherfilename yearbuilt
448  UnivLab_Ashlynn       weather0.csv      2012

Now, use the `read_csv` function to read the following time-series data as Pandas DataFrame: 
1. The meter data file of the identified building (`uid` in the `filtered_metadata`); 
2. The `TemperatureC` and `Humidity` of the identified building (`newweatherfilename` in the `filtered_metadata`);

Tips:
- Be careful about the datatype when getting the values in `filtered_meterdata`
- The weather and meter files are in different folders, and the meter file names need a `.csv` in addition to the name in order to be loaded
- Set the timestamp column as index (`index_col`)
- Have the timestamp as datetime data type (`parse_dates`)
- Keep only the `TemperatureC` and `Humidity` for weather data (`usecols`)

In [ ]:
# Load the specific energy consumption and weather data as identified 
# YOUR CODE HERE


In [ ]:
os.chdir("/content/gdrive/My Drive/EDX Data Science for Construction, Architecture and Engineering/3 - Construction - Pandas Fundamentals/meter_data/")
meterdata = pd.read_csv('UnivLab_Ashlynn.csv', index_col = "timestamp", parse_dates=True) 
meterdata

UnivLab_Ashlynn
timestamp                           
2015-01-01 00:00:00           898.79
2015-01-01 01:00:00           887.98
2015-01-01 02:00:00           892.02
2015-01-01 03:00:00           889.06
2015-01-01 04:00:00           881.37
...                              ...
2015-12-31 19:00:00           869.86
2015-12-31 20:00:00           864.22
2015-12-31 21:00:00           884.05
2015-12-31 22:00:00           857.23
2015-12-31 23:00:00           858.44

[8760 rows x 1 columns]

In [ ]:
os.chdir("/content/gdrive/My Drive/EDX Data Science for Construction, Architecture and Engineering/3 - Construction - Pandas Fundamentals/weather_data/")
weatherdata = pd.read_csv('weather0.csv', index_col = "timestamp", parse_dates=True) 
weatherdata_filtered=weatherdata[['TemperatureC', 'Humidity']]
weatherdata_filtered

TemperatureC  Humidity
timestamp                                  
2015-01-01 00:51:00           2.8      79.0
2015-01-01 01:47:00           3.0      75.0
2015-01-01 01:51:00           2.8      79.0
2015-01-01 02:51:00           2.8      76.0
2015-01-01 03:51:00           2.2      82.0
...                           ...       ...
2015-12-31 19:51:00          12.2      28.0
2015-12-31 20:51:00          11.7      27.0
2015-12-31 21:51:00          10.0      31.0
2015-12-31 22:51:00           9.4      33.0
2015-12-31 23:51:00           7.2      40.0

[8999 rows x 2 columns]

### Exercise 03.1.2 - Resampling Data

As discussed in the videos, very often your IoT data will be at a higher resolution than what you need. Sometimes, the data was not logged at the exact moment as desired and therefore has various timestamps. In a spreadsheet, dealing with these issues can be tricky. With `Pandas`, its much more straightforward.

In this exercise, use the DataFrames of meter and weather data you created to: 
1. create a `all_data` dataframe that holds both the meter data and the weather data at hourly resolution (the resulting dataframe should have a `8760×3` shape)
2. Calculate the weekly **total** energy consumptions and weekly **average** temperature and humidity, save them in a DataFrame called `all_data_weekly` (the resulting dataframe should have a `53×3` shape)

Hint: 
- use `.resample` to unify the timestamp of the two dataframe before `concat` them
- be careful about which axis to `concat`
- the Pandas `resample` function can be used for different basic calculation (e.g. `mean`, `sum`) at different time resolution (e.g. `'H'`, `'M'`)

In [ ]:
# properly resample and concatenate the data
# YOUR CODE HERE


In [ ]:
# Do the weekly calculation
# YOUR CODE HERE


### Exercise 03.1.3 - Plotting

One of the most useful things to do in the early phase of the data analysis process is to visualize the data to find interesting behavior, outliers, etc.

In this exercise, use the Pandas `truncate` and `plot` function to plot the bar chart of weekly temperature and energy consumption of the building during July to September 2015.

Here are some tips that you can use in your code:
- The `truncate` function returns the dataframe between the `before` and `after` date
- The pandas `plot.bar` function returns a `matplotlib.axes.Axes` object, which can be referred to as place to plot
- The temperature and energy data are in different range. Therefore, it is better to have two y axis (`ax`), so that the both trends are clearly shown
- A proper `legend` makes a plot with more than one lines more legible
- Axis ticks of larger `fontsize` are also preferred
- The appearance of the bars can be adjusted by `width` and `position`

In [ ]:
# Process and visualize the data as required
# YOUR CODE HERE


### Please write down the week with the abnormal average temperature (in the form of '2015-XX-XX') as the answer for the quiz question.